sofiAPI.py

```diff
@@ -0,0 +1,634 @@
+            try:
+                username_field = WebDriverWait(driver, 30).until(
+                    EC.element_to_be_clickable((By.XPATH, "//*[@id='username']")))
+                username_field.send_keys(account[0])
+
+                password_field = WebDriverWait(driver, 30).until(
+                    EC.element_to_be_clickable((By.XPATH, "//*[@id='password']")))
+                password_field.send_keys(account[1])
+
+                login_button = WebDriverWait(driver, 30).until(
+                    EC.element_to_be_clickable((By.XPATH, "//*[@id='widget_block']/div/div[2]/div/div/main/section/div/div/div/form/div[2]/button")))
+                login_button.click()
+
+                # Determine if authenticator 2FA is needed
+                secret = account[2] if len(account) > 2 else None
+
+                if secret:
+                    try:
+                        # Use the authenticator 2FA code if the secret exists
+                        two_fa_code = get_2fa_code(secret)
+                        code_field = WebDriverWait(driver, 10).until(
+                            EC.element_to_be_clickable((By.ID, "code")))  # ID for the authenticator code
+                        code_field.send_keys(two_fa_code)
+
+                        code_button = WebDriverWait(driver, 30).until(
+                            EC.element_to_be_clickable((By.CSS_SELECTOR, "#widget_block > div > div.right-column > div > div > main > section > div > div > div > form > div.cc199ae96 > button")))
+                        code_button.click()
+
+                        # Wait for successful login URL
+                        WebDriverWait(driver, 30).until(
+                            EC.url_contains("https://www.sofi.com/member-home/")  # The URL after successful login
+                        )
+
+                    except TimeoutException:
+                        print("Authenticator 2FA code failed or timed out.")
+                        return None  # If authenticator 2FA fails, don't proceed with SMS 2FA
+
+                else:
+                    # Check if the SMS 2FA element is present
+                    try:
+                        WebDriverWait(driver, 5).until(
+                            EC.presence_of_element_located((By.ID, "code"))  # Check if the 2FA SMS code field is present
+                        )
+                        print("SMS 2FA required.")
+                    except TimeoutException:
+                        print("No 2FA required or no SMS 2FA element detected.")
+                        return None
+
+                    # Proceed with SMS 2FA
+                    max_attempts = 3
+                    attempts = 0
+
+                    while attempts < max_attempts:
+                        try:
+                            # Prompt the user for the 2FA code via SMS
+                            code2fa = input('Please enter SMS code and press ENTER to continue:')
```

Comment:
If you check Fidelity, we already have a builtin function for retrieving SMS codes, so use that https://github.com/NelsonDane/auto-rsa/blob/1f93b8702f0f844239aa971df2e54ca1efa127ce/fidelityAPI.py#L163